src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php

```diff
@@ -0,0 +1,219 @@
+            'loop_sleep' => $options['loop_sleep'] ?? ($query['loop_sleep'] ?? 200000),
+            'auto_setup' => $options['auto_setup'] ?? ($query['auto_setup'] ?? true),
+        ];
+    }
+
+    public function publish(string $body, array $headers): void
+    {
+        $sql = $this->driverConnection->createQueryBuilder()
+            ->insert($this->configuration['table_name'])
+            ->values([
+                'body' => ':body',
+                'headers' => ':headers',
+                'queue' => ':queue',
+                'created_at' => ':now',
+                'available_at' => ':now',
+            ]);
+        $now = (new \DateTime('now'))->format('Y-m-d H:i:s');
+        $this->executeQuery($sql, [
+            ':body' => $body,
+            ':headers' => \json_encode($headers),
+            ':queue' => $this->configuration['queue'],
+            ':now' => $now,
+        ]);
+    }
+
+    public function get(): ?array
+    {
+        $this->driverConnection->beginTransaction();
+        try {
+            $query = $this->driverConnection->createQueryBuilder()
+                ->select('m.*')
+                ->from($this->configuration['table_name'], 'm')
+                ->where('m.delivered_at is null')
+                ->andWhere('m.available_at <= :now')
+                ->andWhere('m.queue = :queue')
+                ->orderBy('available_at', 'ASC')
+                ->setMaxResults(1);
+
+            // use SELECT ... FOR UPDATE to lock table
+            $doctrineEnvelope = $this->executeQuery(
+                $query->getSQL().' '.$this->driverConnection->getDatabasePlatform()->getWriteLockSQL(),
+                [
+                    ':now' => (new \DateTime('now'))->format('Y-m-d H:i:s'),
+                    'queue' => $this->configuration['queue'],
+                ]
+            )->fetch();
+
+            if (false === $doctrineEnvelope) {
+                $this->driverConnection->commit();
+
+                return null;
+            }
+
+            $doctrineEnvelope['headers'] = \json_decode($doctrineEnvelope['headers'], true);
+
+            $sql = $this->driverConnection->createQueryBuilder()
+                ->update($this->configuration['table_name'])
+                ->set('delivered_at', ':delivered_at')
+                ->where('id = :id');
+            $this->executeQuery($sql, [
+                ':id' => $doctrineEnvelope['id'],
+                ':delivered_at' => (new \DateTime('now'))->format('Y-m-d H:i:s'),
+            ]);
+
+            $this->driverConnection->commit();
+
+            return $doctrineEnvelope;
+        } catch (\Throwable $e) {
+            $this->driverConnection->rollBack();
+
+            throw $e;
+        }
+    }
+
+    public function ack($id): bool
+    {
+        return $this->driverConnection->delete($this->configuration['table_name'], ['id' => $id]) > 0;
+    }
+
+    public function nack($id): bool
```

Comment:
To match with the `ReceiverInterface` methods:

```suggestion
    public function reject($id): bool
```